### PR TITLE
add literals for sizes in KB, MB, and GB

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@havardpe please review - I want to use this a lot all over the place.  Should 128_MB maybe be 128_KB though?

### CODE REVIEW

**Comment by vekterli** [comment]
The pedant in me would like to comment that these are technically `KiB`, `MiB` and `GiB` since they are powers of two 🤓

**Comment by havardpe** [comment]
> The pedant in me would like to comment that these are technically `KiB`, `MiB` and `GiB` since they are powers of two 🤓

https://en.wikipedia.org/wiki/JEDEC_memory_standards#Unit_prefixes_for_semiconductor_storage_capacity

**Comment by havardpe** [comment]
> The pedant in me would like to comment that these are technically `KiB`, `MiB` and `GiB` since they are powers of two 🤓

If we go down that route, should we also drop the B, since the result is untyped and only really uses the binary scaling prefixes... ?

File: fnet/src/tests/sync_execute/sync_execute.cpp [hunk 1]
```diff
@@ -17,7 +18,7 @@ TEST("sync execute") {
     DoIt exe2;
     DoIt exe3;
     DoIt exe4;
-    FastOS_ThreadPool pool(128 * 1024 * 1024);
+    FastOS_ThreadPool pool(128_MB);
```
Comment (havardpe) [inline fnet/src/tests/sync_execute/sync_execute.cpp:21]: yes, should probably only be 128 KB

File: metrics/src/vespa/metrics/memoryconsumption.cpp [hunk 2]
```diff
@@ -125,12 +126,12 @@ MemoryConsumption::print(std::ostream& out, bool verbose,
 std::string
 MemoryConsumption::bval(uint32_t bytes) {
     std::ostringstream ost;
-    if (bytes < 10 * 1024) {
+    if (bytes < 10_KB) {
         ost << bytes << " B";
-    } else if (bytes < 10 * 1024 * 1024) {
-        ost << (bytes / 1024) << " kB";
+    } else if (bytes < 10_MB) {
+        ost << (bytes / 1_KB) << " kB";
```
Comment (havardpe) [inline metrics/src/vespa/metrics/memoryconsumption.cpp:132]: kB vs KB

**Comment by arnej27959** [comment]
> If we go down that route, should we also drop the B, since the result is without unit and only really uses the binary scaling prefixes... ?

that may actually be better; in most cases we're talking about (memory) size in bytes, but there are some exceptions like:
`constexpr uint32_t DEFAULT_MAX_LIDS_PER_FILE = 32 * 1024 * 1024;`
and we could write the number as `32_Mi` instead.

**Comment by arnej27959** [comment]
several options here, please vote for your favourite :-)

File: vespalib/src/tests/alloc/alloc_test.cpp [hunk 3]
```diff
@@ -215,15 +216,15 @@ TEST("mmap alloc can be extended if room") {
     Alloc buf = Alloc::allocMMap(100);
 
     TEST_DO(ensureRoomForExtension(buf, reserved));
-    TEST_DO(verifyExtension(buf, 4096, 4096*2));
+    TEST_DO(verifyExtension(buf, 4_KB, 4_KB*2));
```
Comment (havardpe) [inline vespalib/src/tests/alloc/alloc_test.cpp:219]: 8_KB?

File: vespalib/src/vespa/vespalib/util/size_literals.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+#pragma once
+
+#include <cstddef>
+
+constexpr size_t operator "" _KB(unsigned long long kb_in) {
+	return size_t(kb_in << 10u);
+}
+
+constexpr size_t operator "" _MB(unsigned long long kb_in) {
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/util/size_literals.h:11]: param name

File: vespalib/src/vespa/vespalib/util/size_literals.h [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+	return size_t(kb_in << 20u);
+}
+
+constexpr size_t operator "" _GB(unsigned long long kb_in) {
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/util/size_literals.h:15]: param name

**Review by havardpe — APPROVED** [review]
consider adding a small test

I personally prefer KB/MB/GB, but will also be ok with the more correct KiB/MiB/GiB (or even Ki/Mi/Gi)

**Comment by havardpe** [comment]
we could also type the result from KB/KiB like chrono does, but that might be over-complicating things

**Comment by baldersheim** [comment]
I like just K M G and T

**Comment by vekterli** [comment]
I'm now leaning towards `Ki`, `Mi` etc.

_Pistols at dawn._

**Comment by toregge** [comment]
I vote for `Ki`, `Mi`, `Gi`, `Ti`, `Pi`, `Ei`, `Zi`, `Yi`
("Specific units of IEC 60027-2 A.2 and ISO/IEC 80000")

**Comment by geirst** [comment]
I vote for Ki/Mi/Gi.

**Comment by arnej27959** [comment]
changed to _Ki etc and used them more places.

please take a quick scan to see if any changes look wrong.

File: vespalib/src/vespa/vespalib/net/tls/protocol_snooping.cpp [hunk 6]
```diff
@@ -63,14 +64,14 @@ TlsSnoopingResult snoop_client_hello_header(const char* buf) noexcept {
...
-    // Length of TLS record follows. Must be <= 16KiB + 2048 (16KiB + 256 on v1.3).
+    // Length of TLS record follows. Must be <= 16KiB + 2_Ki (16KiB + 256 on v1.3).
     // We expect that the first record contains _only_ a ClientHello with no coalescing
     // and no fragmentation. This is technically a violation of the TLS spec, but this
     // particular detection logic is only intended to be used against other Vespa nodes
     // where we control frame sizes and where such fragmentation should not take place.
     // We also do not support TLSv1.3 0-RTT which may trigger early data.
     uint16_t length = tls_record_length(buf);
-    if ((length < 4) || (length > (16384 + 2048))) {
+    if ((length < 4) || (length > (16_Ki + 2048))) {
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/net/tls/protocol_snooping.cpp:74]: Consider _16_Ki + 2_Ki_.

File: vespalib/src/vespa/vespalib/util/alloc.cpp [hunk 7]
```diff
@@ -233,7 +234,7 @@ createAutoAllocatorsWithDefault() {
 AutoAllocatorsMapWithDefault  _G_availableAutoAllocators = createAutoAllocatorsWithDefault();
 alloc::HeapAllocator _G_heapAllocatorDefault;
 alloc::AlignedHeapAllocator _G_4KalignedHeapAllocator(1024);
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/util/alloc.cpp:236]: Consider _1_Ki_.

File: searchcore/src/tests/proton/index/indexmanager_test.cpp [hunk 8]
```diff
@@ -316,7 +317,7 @@ TEST_F(IndexManagerTest, require_that_memory_index_is_flushed)
 
 TEST_F(IndexManagerTest, require_that_large_memory_footprint_triggers_urgent_flush) {
     using FlushStats = IndexMaintainer::FlushStats;
-    constexpr size_t G = 1024*1024*1024l;
+    constexpr size_t G =1_Gi;
```
Comment (geirst) [inline searchcore/src/tests/proton/index/indexmanager_test.cpp:320]: Consider 15_Gi and 17_Gi further down instead.

File: searchcore/src/tests/proton/server/memory_flush_config_updater/memory_flush_config_updater_test.cpp [hunk 9]
```diff
@@ -83,7 +83,7 @@ TEST("require that we use configured memory limits") {
 
 TEST("require that we cap configured limits based on available memory") {
     const uint64_t LIMIT = defaultMemory.sizeBytes()/4;
-    constexpr uint64_t MEM_4G = 4 * ONE_G;
+    constexpr uint64_t MEM_4G = 4_Gi;
```
Comment (geirst) [inline searchcore/src/tests/proton/server/memory_flush_config_updater/memory_flush_config_updater_test.cpp:86]: Consider using 4_Gi directly on the next line instead.

File: searchcorespi/src/vespa/searchcorespi/index/indexflushtarget.cpp [hunk 10]
```diff
@@ -41,7 +42,7 @@ IndexFlushTarget::needUrgentFlush() const
 {
     // Due to limitation of 16G address space of single datastore
     // TODO: Even better if urgency was decided by memory index itself.
-    constexpr int64_t G = 1024*1024*1024l;
+    constexpr int64_t G =1_Gi;
```
Comment (geirst) [inline searchcorespi/src/vespa/searchcorespi/index/indexflushtarget.cpp:45]: Consider using 16_Gi further down instead.

File: storageserver/src/tests/testhelper.cpp [hunk 11]
```diff
@@ -50,7 +50,7 @@ vdstestlib::DirConfig getStandardConfig(bool storagenode) {
     // Easier to see what goes wrong with only 1 thread per disk.
     config->set("minimum_file_meta_slots", "2");
     config->set("minimum_file_header_block_size", "368");
-    config->set("minimum_file_size", "4096");
+    config->set("minimum_file_size", "4_Ki");
```
Comment (geirst) [inline storageserver/src/tests/testhelper.cpp:53]: What will be the actual string here? If "4_Ki", I guess we must change to "4096".

File: storageserver/src/tests/testhelper.cpp [hunk 12]
```diff
@@ -66,7 +66,7 @@ vdstestlib::DirConfig getStandardConfig(bool storagenode) {
...
```
Comment (geirst) [inline storageserver/src/tests/testhelper.cpp:69]: What will be the actual string here? If "4_Ki", I guess we must change to "4096".

File: vespalib/src/vespa/vespalib/util/size_literals.h [hunk 13]
```diff
@@ -0,0 +1,21 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#pragma once
+
+#include <cstddef>
+
+constexpr size_t operator "" _Ki(unsigned long long k_in) {
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/util/size_literals.h:7]: Consider adding a unit tests for the literals as well.

**Review by geirst — COMMENTED** [review]
This is great 🥇 . Some comments to consider.

**Review by geirst — APPROVED** [review]
👍

**Review by vekterli — APPROVED** [review]
:+1:

**Review by toregge — APPROVED** [review]

**Pull request merged**